doc/design/to_run.md

```diff
@@ -0,0 +1,165 @@
+# SQLFlow Syntax Extension: `TO RUN`
+
+
+## Data Transformation
+
+As the purpose of SQLFlow is to extend SQL syntax for end-to-end machine
+learning, we have to consider *data transformation*.  Usually, there are two
+kinds of data transformations:
+
+1. *pre-processing*, which happens before the training stage. Examples include
+   joining a feature lookup table with the training data table.
+1. *in-train loop transformation*, which happens in the training loop for each
+   data instance.
+
+The in-train loop transformation differs from the pre-processing primarily in
+that they are part of the trained models and used in predictions.
+
+SQLFlow users can describe the in-train loop transformation using the `COLUMN`
+clause in the `TO TRAIN`-clause, and they can write standard statements like
+`SELECT .. INTO new_table` to pre-process data.
+
+This design is about pre-processing that we cannot express using standard SQL
+statements.
+
+
+## The Challenge
+
+SQL is extensible and thus powerful.  Most DBMS support user-defined functions
+(UDFs).  The SELECT statement applies UDFs and built-in functions to each row of
+a table/view.  However, some pre-processing transformations are not row-wise.
+An example is *sliding-window*.
+
+A [user told us](https://github.com/sql-machine-learning/sqlflow/issues/2238)
+that he wants to extract temporal features from time-series data by calling the
+well-known Python package [`tsfresh`](https://tsfresh.readthedocs.io), which
+runs a sliding-window over the time-series data and convert each step into a new
+row.
+
+This requirement intrigues us to support a general way to call arbitrary
+programs from SQL, hence the idea of the `TO RUN` clause.
+
+
+## Syntax Extension
+
+The subject to run is a program.  A program needs versioning and releasing.  As
+always, we assume Docker images are released form.  Therefore, the user needs to
+provide a Docker image as the subject of the `TO RUN` clause.  The SQLFlow
+compiler translates the TO RUN statement into a Tekton step that runs this
+Docker image.
+
+```sql
+SELECT * FROM source_table ORDER BY creation_date
+TO RUN 
+  a_data_scientist/extract_ts_features:1.0
+CMD 
+  "--verbose",
+  "--window_width=120"
+  "output_table_name";
```

Comment:
From the discussion, the `output_table_name` is in `INTO` clause.

```SQL
SELECT * FROM source_table ORDER BY creation_date
TO RUN 
  a_data_scientist/extract_ts_features:1.0
CMD 
  "--verbose",
  "--window_width=120"
INTO output_table_1, output_table_2
```

Reply:
You are right. The names after INTO are supposed to be passed to the entrypoint program via environment variables.